lib/services/role.go

```diff
@@ -1163,7 +1116,11 @@ func (set RoleSet) GetAllowedLoginsForResource(resource AccessCheckable) ([]stri
 		}
 
 		for _, login := range loginGetter(types.Allow) {
-			mapped[login] = true
+			// Only set to true if not already set, the login is denied if any
+			// role denies it.
+			if _, alreadySet := mapped[login]; !alreadySet {
+				mapped[login] = true
+			}
```

Comment:
It's not related to the label expression at all, I just saw the logic that looks flawed and tried to improve it. This loop was not really deterministic before, if one role allows the login and another denies it, it would depend on the order.

Technically it's not a bug because at the end of this function we do another pass of all the `notDenied` logins and check access to them again, but we are doing that more times than necessary when we already know the login is denied.